### PR TITLE
Update errorhandling for Wallets, SIOP QR Codes

### PR DESCRIPTION
**Pull request by nodh**
Implements improvements for Wallet Apps:

- Passes exceptions from native crypto service implementations up to callers of the `JwsService`
- Adds QR Code use case for SIOPv2
- Store entries are serializable

In my opinion we should release version 3.4.0 afterwards.

### CODE REVIEW

File: vclib-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopVerifier.kt [hunk 1]
```diff
@@ -95,6 +95,48 @@ class OidcSiopVerifier(
         )
     }
 
+    private val containerJwt = FormatContainerJwt(algorithms = arrayOf(JwsAlgorithm.ES256.identifier))
```
Comment (JesusMcCloud) [inline vclib-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopVerifier.kt:98]: Why is the algo hardcoded here?

File: vclib-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopWallet.kt [hunk 2]
```diff
@@ -231,9 +235,10 @@ class OidcSiopWallet(
         )
         val jwsPayload = idToken.serialize().encodeToByteArray()
         val jwsHeader = JwsHeader(algorithm = JwsAlgorithm.ES256)
```
Comment (JesusMcCloud) [inline vclib-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopWallet.kt:237]: should we not take this opportunity to get rid of all the hardcoded stuff and at least make it a constructor property, that defaults to ES256? I'm not saying this PR should do it, but before we release 3.4.0, those hardcoded values should go.

**Review by JesusMcCloud — COMMENTED** [review]
Looks good except for the remnants of hardcoded P-256-based algos. But that's a separate issue and should be done in a separate PR (but still before 3.4.0, becuase now we have a CryptoService that supports all curves, but other pieces of the puzzle are hardcoded to only fit P-256.

**Review by JesusMcCloud — APPROVED** [review]
Looks good except for the remnants of hardcoded P-256-based algos. But that's a separate issue and should be done in a separate PR (but still before 3.4.0, becuase now we have a CryptoService that supports all curves, but other pieces of the puzzle are hardcoded to only fit P-256.

File: vclib-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopWallet.kt (hunk 2 above)
Comment (nodh) [reply vclib-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopWallet.kt:237]: But in `JwsHeader`, the `algorithm` parameter is required, i.e. non-null.

File: vclib-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopVerifier.kt (hunk 1 above)
Comment (nodh) [reply vclib-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopVerifier.kt]: Fixed that in recent commits, for several usages of hardcoded algorithms in the OID4VCI/SIOP implementations.

File: vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/jws/JwsService.kt [hunk 3]
```diff
@@ -107,36 +126,33 @@ class DefaultJwsService(private val cryptoService: CryptoService) : JwsService {
...
+        val enc = header.encryption
+            ?: return KmmResult.failure(IllegalArgumentException("No encryption in JWE header"))
+        val epk = header.ephemeralKeyPair
+            ?: return KmmResult.failure(IllegalArgumentException("No epk in JWE header"))
+        val z = cryptoService.performKeyAgreement(epk, alg).getOrElse {
+            Napier.w("No Z value from native code", it)
+            return KmmResult.failure(it)
+        }
+        val kdfInput = prependWithAdditionalInfo(z, enc, header.agreementPartyUInfo, header.agreementPartyVInfo)
         val key = cryptoService.messageDigest(kdfInput, Digest.SHA256).getOrElse {
```
Comment (JesusMcCloud) [inline vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/jws/JwsService.kt:143]: 🤔

File: vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/jws/JwsService.kt [hunk 4]
```diff
@@ -165,23 +180,22 @@ class DefaultJwsService(private val cryptoService: CryptoService) : JwsService {
         val z = cryptoService.performKeyAgreement(ephemeralKeyPair, recipientKey, jweAlgorithm)
             .getOrElse {
                 Napier.w("No Z value from native code", it)
-                return null
+                return KmmResult.failure(it)
             }
         val kdf = prependWithAdditionalInfo(z, jweEncryption, null, null)
         val key = cryptoService.messageDigest(kdf, Digest.SHA256).getOrElse {
```
Comment (JesusMcCloud) [inline vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/jws/JwsService.kt:186]: 🤔

**Review by JesusMcCloud — COMMENTED** [review]
Are the hardcoded SHA-256 digests a problem?

File: vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/jws/JwsService.kt (hunk 4 above)
Comment (nodh) [reply vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/jws/JwsService.kt:186]: This is fixed per spec https://datatracker.ietf.org/doc/html/rfc7518#section-4.6.2

**Review by JesusMcCloud — APPROVED** [review]

**Pull request merged**